src/lib/input/input-container.ts

```diff
@@ -311,9 +295,8 @@ export class MdInputDirective implements OnChanges, OnDestroy, DoCheck {
   /** Re-evaluates the error state. This is only relevant with @angular/forms. */
   private _updateErrorState() {
     const oldState = this._isErrorState;
-    const control = this._ngControl;
-    const parent = this._parentFormGroup || this._parentForm;
-    const newState = control && this.errorStateMatcher(control.control as FormControl, parent);
+    const errorMatcher = this.errorStateMatcher || this._errorOptions.errorStateMatcher;
```

Comment:
Alright, let's just go back to the injection token with a plain function and we can worry about a provider if we really need it.